webserver/webserver/src/main/java/io/helidon/webserver/HttpInitializer.java

```diff
@@ -72,17 +89,53 @@
         this.webServer = webServer;
     }
 
+    /**
+     * Calls release on every ReferenceHoldingQueue that has been deemed ready for
+     * garbage collection and added to {@code queues}. The ReferenceHoldingQueue's not
+     * fully released are added to {@code unreleasedQueues} for later retries. Uses
+     * a lock to avoid concurrent modifications to {@code queues}.
+     */
+    @SuppressWarnings("unchecked")
     private void clearQueues() {
-        queues.removeIf(ReferenceHoldingQueue::release);
+        if (clearLock.get() || !clearLock.compareAndSet(false, true)) {
+            return;
+        }
+        try {
+            for (Reference<?> r = queues.poll(); r != null; r = queues.poll()) {
+                if (!(r instanceof IndirectReference<?, ?>)) {
+                    LOGGER.finest("Unexpected reference in queues");
+                    continue;
+                }
+                ReferenceHoldingQueue<?> q = ((IndirectReference<?, ReferenceHoldingQueue<?>>) r).acquire();
+                if (q == null) {
+                    continue;       // no longer referenced
+                }
+                if (!q.release()) {
+                    unreleasedQueues.add(q);
+                }
+            }
+            unreleasedQueues.removeIf(ReferenceHoldingQueue::release);
+        } finally {
+            clearLock.lazySet(false);
```

Comment:
Just being careful: `lazySet` is designed for very specialized use cases.  Are you sure `set` isn't the right choice here?

Reply:
Yes, this is one case where it makes sense.

We only have a `lock` that can either be taken immediately, or the contender goes away:
https://github.com/oracle/helidon/pull/2704/files/06c459f90972ba6a14a500bfaea7cd4b91beb3cb#diff-a0cac19f24ae85fbc99afd8c0b4e8375d3c036a0ad949c68befe56fb7346461dR100 - that yellow line

In this case we only need to ensure the `get()` and `compareAndSet` on that line synchronize-with this `lazySet`, which they do.

We don't need to "publish" any other changes, as queue poll and removeIf are atomic in their own right.